File: Solutions/860-design-circular-queue/design-circular-queue.cs
using System;

/// <summary>
/// Implementation of a circular queue (ring buffer) using an array.
/// Supports O(1) enqueue, dequeue, front, rear, isEmpty, isFull operations.
/// </summary>
public class MyCircularQueue
{
    private int[] queue;    // Array to store elements
    private int front;      // Index of the front element
    private int rear;       // Index where the next element will be inserted
    private int size;       // Current number of elements
    private int capacity;   // Maximum capacity of the queue

    /// <summary>
    /// Initialize the queue with capacity k.
    /// </summary>
    public MyCircularQueue(int k)
    {
        capacity = k;               // Set maximum capacity
        queue = new int[k];         // Allocate array
        front = 0;                  // Front starts at 0
        rear = 0;                   // Rear also starts at 0
        size = 0;                   // Initially, queue is empty
    }

    /// <summary>
    /// Inserts an element into the circular queue.
    /// Returns true if the operation is successful.
    /// </summary>
    public bool EnQueue(int value)
    {
        if (IsFull()) return false;    // Cannot insert if queue is full

        queue[rear] = value;           // Place value at rear index
        rear = (rear + 1) % capacity;  // Move rear pointer in circular manner
        size++;                         // Increase the size
        return true;
    }

    /// <summary>
    /// Deletes an element from the circular queue.
    /// Returns true if the operation is successful.
    /// </summary>
    public bool DeQueue()
    {
        if (IsEmpty()) return false;    // Cannot dequeue if queue is empty

        front = (front + 1) % capacity; // Move front pointer in circular manner
        size--;                          // Decrease the size
        return true;
    }

    /// <summary>
    /// Gets the front item from the queue. Returns -1 if empty.
    /// </summary>
    public int Front()
    {
        if (IsEmpty()) return -1;
        return queue[front];            // Return element at front index
    }

    /// <summary>
    /// Gets the last item (rear) from the queue. Returns -1 if empty.
    /// </summary>
    public int Rear()
    {
        if (IsEmpty()) return -1;
        // Rear points to next insertion index, so rear element is at (rear - 1 + capacity) % capacity
        return queue[(rear - 1 + capacity) % capacity];
    }

    /// <summary>
    /// Checks whether the circular queue is empty.
    /// </summary>
    public bool IsEmpty()
    {
        return size == 0;
    }

    /// <summary>
    /// Checks whether the circular queue is full.
    /// </summary>
    public bool IsFull()
    {
        return size == capacity;
    }
}

/**
 * Example Usage:
 * MyCircularQueue circularQueue = new MyCircularQueue(3);
 * circularQueue.EnQueue(1); // true
 * circularQueue.EnQueue(2); // true
 * circularQueue.EnQueue(3); // true
 * circularQueue.EnQueue(4); // false (queue is full)
 * circularQueue.Rear();     // 3
 * circularQueue.IsFull();   // true
 * circularQueue.DeQueue();  // true
 * circularQueue.EnQueue(4); // true
 * circularQueue.Rear();     // 4
 */

File: Solutions/860-design-circular-queue/design-circular-queue.swift
class MyCircularQueue {

    // Array used to store queue elements
    private var queue: [Int]

    // Maximum size of the circular queue
    private var capacity: Int

    // Index pointing to the front element
    private var front: Int

    // Index pointing to the position
    // where the next element will be inserted
    private var rear: Int

    // Current number of elements in the queue
    private var count: Int

    // Initialize the circular queue with size k
    init(_ k: Int) {

        // Set the capacity of the queue
        self.capacity = k

        // Initialize the array with k elements
        // Default value does not matter as count controls validity
        self.queue = Array(repeating: 0, count: k)

        // Front starts at index 0
        self.front = 0

        // Rear also starts at index 0
        self.rear = 0

        // Initially, the queue is empty
        self.count = 0
    }

    // Insert an element into the circular queue
    func enQueue(_ value: Int) -> Bool {

        // If the queue is already full, insertion fails
        if isFull() {
            return false
        }

        // Insert the value at the rear position
        queue[rear] = value

        // Move rear forward circularly
        rear = (rear + 1) % capacity

        // Increase the number of elements in the queue
        count += 1

        // Insertion was successful
        return true
    }

    // Delete an element from the circular queue
    func deQueue() -> Bool {

        // If the queue is empty, deletion fails
        if isEmpty() {
            return false
        }

        // Move front forward circularly
        front = (front + 1) % capacity

        // Decrease the number of elements in the queue
        count -= 1

        // Deletion was successful
        return true
    }

    // Get the front element of the queue
    func Front() -> Int {

        // If the queue is empty, return -1
        if isEmpty() {
            return -1
        }

        // Return the front element
        return queue[front]
    }

    // Get the last element of the queue
    func Rear() -> Int {

        // If the queue is empty, return -1
        if isEmpty() {
            return -1
        }

        // Calculate the index of the last element
        let lastIndex = (rear - 1 + capacity) % capacity

        // Return the last element
        return queue[lastIndex]
    }

    // Check if the queue is empty
    func isEmpty() -> Bool {

        // Queue is empty when count is zero
        return count == 0
    }

    // Check if the queue is full
    func isFull() -> Bool {

        // Queue is full when count equals capacity
        return count == capacity
    }
}


/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * let obj = MyCircularQueue(k)
 * let ret_1: Bool = obj.enQueue(value)
 * let ret_2: Bool = obj.deQueue()
 * let ret_3: Int = obj.Front()
 * let ret_4: Int = obj.Rear()
 * let ret_5: Bool = obj.isEmpty()
 * let ret_6: Bool = obj.isFull()
 */